Light and dark themes

Hi @echarles, In previous discussions, we touched upon the topic of themes. I would like to inquire about the current status of theme development and if there are any workarounds to enable themes before they are officially incorporated into the library. I attempted the approaches discussed in this thread (https://github.com/datalayer/jupyter-ui/issues/85), but so far not working. Do you have any insights or suggestions that could help? Thank you in advance!

Hi @echarles, I included the webpack rules and installed styled-loader and css-loader, and see this error now: 
```
 ⨯ ./node_modules/@datalayer/jupyter-react/node_modules/@jupyterlab/cells/node_modules/@jupyterlab/codemirror/style/base.css:5:0
Module not found: Invalid generator object. Asset Modules Plugin has been initialized using a generator object that does not match the API schema.
 - generator has an unknown property 'filename'. These properties are valid:
   object { dataUrl? }
   -> Generator options for asset/inline modules.

https://nextjs.org/docs/messages/module-not-found
```

--- a/examples/next-js/src/app/page.tsx
+++ b/examples/next-js/src/app/page.tsx
@@ -7,7 +7,10 @@
 'use client'
 
 import { useState, useEffect, useCallback } from 'react';
-import { Text, ToggleSwitch } from '@primer/react';
+import { theme as primerTheme, Text, ToggleSwitch, Box } from '@primer/react';
+import { Theme } from '@primer/react/lib/ThemeProvider';
+import { ColorMode } from '@datalayer/jupyter-react/lib/jupyter/lab/JupyterLabColorMode';
+import { jupyterTheme } from '@datalayer/jupyter-react/lib/jupyter/theme';
 import dynamic from 'next/dynamic';
 import { useTheme, ThemeProvider } from "next-themes";
 
@@ -23,48 +26,89 @@ const CellComponentNoSSR = dynamic(
 
 const JupyterDemo = () => {
 
-  const { themes, theme, setTheme } = useTheme();
-  console.log('---', themes)
+  // next theme can be 'light', 'dark' or 'system'.
+  const { theme: nextColorMode, setTheme: setNextColorMode } = useTheme();
 
+  const [colorMode, setColorMode] = useState<ColorMode>('light');
   const [isOn, setIsOn] = useState(false);
+
+  const [theme, setTheme] = useState<Theme>(jupyterTheme);
+  const [isThemeOn, setIsThemeOn] = useState(false);
 
   useEffect(() => {
     if (isOn) {
-      setTheme('dark');
+      setColorMode('dark');
+      setNextColorMode('dark');
     } else {
-      setTheme('light');
+      setColorMode('light');
+      setNextColorMode('light');
     }
   }, [isOn]);
-
   const onClick = useCallback(() => {
     setIsOn(!isOn);
   }, [isOn]);
-
   const handleSwitchChange = useCallback((on: boolean) => {
     setIsOn(on);
   }, []);
 
+  useEffect(() => {
+    if (isThemeOn) {
+      setTheme(primerTheme);
+    } else {
+      setTheme(jupyterTheme);
+    }
+  }, [isThemeOn]);
+  const onThemeClick = useCallback(() => {
+    setIsThemeOn(!isThemeOn);
+  }, [isThemeOn]);
+  const handleThemeSwitchChange = useCallback((on: boolean) => {
+    setIsThemeOn(on);
+  }, []);
+
   return (
     <>
-     <Text
-        fontSize={2}
-        fontWeight="bold"
-        id="switch-label"
-        display="block"
-        mb={1}
-      >
-        { theme === 'light' ? 'Light' : 'Dark' } Mode
-      </Text>
-      <ToggleSwitch
-        size="small"
-        onClick={onClick}
-        onChange={handleSwitchChange}
-        checked={isOn}
-        statusLabelPosition="end"
-        aria-labelledby="switch-label"
-      />
-      <NotebookComponentNoSSR colorMode={theme === 'light' ? 'light' : 'dark'} />
-      <CellComponentNoSSR colorMode={theme === 'light' ? 'light' : 'dark'} />
+      <Box display="flex">
+        <Box mr={3}>
+          <Text
+            fontSize={2}
+            fontWeight="bold"
+            id="switch-label"
+            display="block"
+            mb={1}
+          >
+            { colorMode === 'light' ? 'Light' : 'Dark' } Mode
+          </Text>
+          <ToggleSwitch
+            size="small"
+            onClick={onClick}
+            onChange={handleSwitchChange}
+            checked={isOn}
+            statusLabelPosition="end"
+            aria-labelledby="switch-label-color-mode"
+          />
+        </Box>
+        <Box>
+          <Text
+          fontSize={2}
+          fontWeight="bold"
+          id="switch-label"
+          display="block"
+          mb={1}
+          >
+            Primer Theme
+          </Text>
+          <ToggleSwitch
+            size="small"
+            onClick={onThemeClick}
+            onChange={handleThemeSwitchChange}
+            checked={isThemeOn}
+            statusLabelPosition="end"
+            aria-labelledby="switch-label-theme"
+          />
+        </Box>
+      </Box>
+      <NotebookComponentNoSSR colorMode={colorMode} theme={theme}/>
+      <CellComponentNoSSR colorMode={colorMode} theme={theme}/>
     </>
   )
 }

--- a/examples/next-js/src/components/CellComponent.tsx
+++ b/examples/next-js/src/components/CellComponent.tsx
@@ -7,13 +7,15 @@
 'use client'
 
 import { Jupyter, Cell } from '@datalayer/jupyter-react';
+import { Theme } from '@primer/react/lib/ThemeProvider';
 
 type CellComponentProps = {
   colorMode: 'light' | 'dark';
+  theme: Theme;
 }
 
 export const CellComponent = (props: CellComponentProps) => {
-  const { colorMode } = props;
+  const { colorMode, theme } = props;
   return (
     <>
       <div style={{fontSize: 20}}>Jupyter Cell in Next.js</div>
@@ -22,6 +24,7 @@ export const CellComponent = (props: CellComponentProps) => {
         jupyterServerWsUrl="wss://oss.datalayer.tech/api/kernel"
         jupyterToken="60c1661cc408f978c309d04157af55c9588ff9557c9380e4fb50785750703da6"
         colorMode={colorMode}
+        theme={theme}
       >
         <Cell/>
     </Jupyter>

--- a/examples/next-js/src/components/NotebookComponent.tsx
+++ b/examples/next-js/src/components/NotebookComponent.tsx
@@ -7,13 +7,16 @@
 'use client'
 
 import { Jupyter, Notebook, CellSidebar } from '@datalayer/jupyter-react';
+import { NotebookToolbar } from '@datalayer/jupyter-react/lib/examples/toolbars/NotebookToolbar';
+import { Theme } from '@primer/react/lib/ThemeProvider';
 
 type NotebookComponentProps = {
   colorMode: 'light' | 'dark';
+  theme: Theme;
 }
 
 export const NotebookComponent = (props: NotebookComponentProps) => {
-  const { colorMode } = props;
+  const { colorMode, theme } = props;
   return (
     <>
       <div style={{fontSize: 20}}>Jupyter Notebook in Next.js</div>
@@ -22,12 +25,14 @@ export const NotebookComponent = (props: NotebookComponentProps) => {
         jupyterServerWsUrl="wss://oss.datalayer.tech/api/kernel"
         jupyterToken="60c1661cc408f978c309d04157af55c9588ff9557c9380e4fb50785750703da6"
         colorMode={colorMode}
+        theme={theme}
       >
         <Notebook
           path="ipywidgets.ipynb"
           uid="notebook-nextjs-1"
           cellSidebarMargin={120}
           CellSidebar={CellSidebar}
+          Toolbar={NotebookToolbar}
         />
     </Jupyter>
   </>

--- a/packages/react/src/examples/NotebookThemeColorMode.tsx
+++ b/packages/react/src/examples/NotebookThemeColorMode.tsx
@@ -1,0 +1,114 @@
+/*
+ * Copyright (c) 2021-2023 Datalayer, Inc.
+ *
+ * MIT License
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { createRoot } from 'react-dom/client';
+import { Text, ToggleSwitch, theme as primerTheme, Box } from '@primer/react';
+import { Theme } from '@primer/react/lib/ThemeProvider';
+import { INotebookContent } from '@jupyterlab/nbformat';
+import { ColorMode } from '../jupyter/lab/JupyterLabColorMode';
+import Jupyter from '../jupyter/Jupyter';
+import { jupyterTheme } from '../jupyter/theme';
+import Notebook from '../components/notebook/Notebook';
+import NotebookToolbar from './toolbars/NotebookToolbar';
+import CellSidebar from '../components/notebook/cell/sidebar/CellSidebar';
+
+import nbformat from './notebooks/NotebookExample1.ipynb.json';
+
+const NotebookThemeColorMode = () => {
+  const [theme, setTheme] = useState<Theme>(jupyterTheme);
+  const [isThemeOn, setIsThemeOn] = useState(false);
+  const [colorMode, setColorMode] = useState<ColorMode>('light');
+  const [isOn, setIsOn] = useState(false);
+
+  useEffect(() => {
+    if (isThemeOn) {
+      setTheme(primerTheme);
+    } else {
+      setTheme(jupyterTheme);
+    }
+  }, [isThemeOn]);
+  const onThemeClick = useCallback(() => {
+    setIsThemeOn(!isThemeOn);
+  }, [isThemeOn]);
+  const handleThemeSwitchChange = useCallback((on: boolean) => {
+    setIsThemeOn(on);
+  }, []);
+
+  useEffect(() => {
+    if (isOn) {
+      setColorMode('dark');
+    } else {
+      setColorMode('light');
+    }
+  }, [isOn]);
+  const onClick = useCallback(() => {
+    setIsOn(!isOn);
+  }, [isOn]);
+  const handleSwitchChange = useCallback((on: boolean) => {
+    setIsOn(on);
+  }, []);
+  return (
+    <>
+      <Jupyter theme={theme} colorMode={colorMode}>
+        <Box display="flex">
+          <Box mr={3}>
+            <Text
+            fontSize={2}
+            fontWeight="bold"
+            id="switch-label"
+            display="block"
+            mb={1}
+            >
+              Primer Theme
+            </Text>
+            <ToggleSwitch
+              size="small"
+              onClick={onThemeClick}
+              onChange={handleThemeSwitchChange}
+              checked={isThemeOn}
+              statusLabelPosition="end"
+              aria-labelledby="switch-label-theme"
+            />
+          </Box>
+          <Box>
+            <Text
+              fontSize={2}
+              fontWeight="bold"
+              id="switch-label"
+              display="block"
+              mb={1}
+            >
+              { colorMode === 'light' ? 'Light' : 'Dark' } Mode
+            </Text>
+            <ToggleSwitch
+              size="small"
+              onClick={onClick}
+              onChange={handleSwitchChange}
+              checked={isOn}
+              statusLabelPosition="end"
+              aria-labelledby="switch-label-color-mode"
+            />
+          </Box>
+        </Box>
+        <Notebook
+          nbformat={nbformat as INotebookContent}
+          uid="notebook-model-uid"
+          height="calc(100vh - 2.6rem)" // (Height - Toolbar Height).
+          cellSidebarMargin={120}
+          CellSidebar={CellSidebar}
+          Toolbar={NotebookToolbar}
+        />
+      </Jupyter>
+    </>
+  );
+};
+
+const div = document.createElement('div');
+document.body.appendChild(div);
+const root = createRoot(div);
+
+root.render(<NotebookThemeColorMode />);

--- a/packages/react/src/jupyter/lab/JupyterLabCss.tsx
+++ b/packages/react/src/jupyter/lab/JupyterLabCss.tsx
@@ -55,11 +55,11 @@ export const JupyterLabCss = (props: Props) => {
     let theme;
     switch (colorMode) {
       case 'light': {
-        theme = import('@jupyterlab/theme-light-extension/style/theme.css');
+        theme = import('!!raw-loader!@jupyterlab/theme-light-extension/style/variables.css');
         break;
       }
       case 'dark': {
-        theme = import('@jupyterlab/theme-dark-extension/style/theme.css');
+        theme = import('!!raw-loader!@jupyterlab/theme-dark-extension/style/variables.css');
         break;
       }
     }
@@ -69,6 +69,7 @@ export const JupyterLabCss = (props: Props) => {
     //   with css-loader as string - this is available in css-loader v6.3.0 or later:
     //   { test: /style\/theme\.css$/i, loader: 'css-loader', options: {exportType: 'string'} }
     theme?.then(module => {
+
       if (module.default) {
         document.body.insertAdjacentHTML(
           'afterbegin',

--- a/packages/react/src/typings.d.ts
+++ b/packages/react/src/typings.d.ts
@@ -34,6 +34,11 @@ declare module '*/style/index.js' {
   export default value;
 }
 
+declare module '*/variables.css' {
+  const value: any;
+  export default value;
+}
+
 declare module '*/theme.css' {
   const value: any;
   export default value;

--- a/packages/react/webpack.config.js
+++ b/packages/react/webpack.config.js
@@ -43,6 +43,7 @@ const ENTRY =
   // './src/examples/NotebookNbFormatChange';
   // './src/examples/NotebookSkeleton';
   // './src/examples/NotebookTheme';
+  // './src/examples/NotebookThemeColorMode';
   // './src/examples/ObservableHQ';
   // './src/examples/Outputs';
   // './src/examples/Plotly';

